crates/ecs/src/lib.rs

```diff
@@ -585,27 +978,46 @@ impl<'a, Component> Deref for Read<'a, Component> {
     }
 }
 
-impl<Component: Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
-    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
-
-    fn borrow(world: &World) -> Self::BorrowedData<'_> {
-        world.borrow_component_vec::<Component>()
+impl<Component: Debug + Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
+    type BorrowedData<'components> = (
+        BorrowedArchetypeIdx,
+        Vec<(ComponentIdx, ReadComponentVec<'components, Component>)>,
+    );
+
+    fn borrow<'a>(
+        world: &'a World,
+        signature: &[TypeId],
+    ) -> SystemParameterResult<Self::BorrowedData<'a>> {
+        let component_vecs = world
+            .borrow_component_vecs_with_signature::<Component>(signature)
+            .map_err(SystemParameterError::BorrowComponentVecs)?;
```

Comment:
There should not be a runtime error when a system is querying a component type that has not been added to any entity. The component may be added later (when it is possible to add/remove components in runtime).

Reply:
I was also thinking about this. In this case this error is only thrown if no archetypes exists of associated with a type. But this should only result in no archetypes being returned